### ISSUE TITLE
Ported query consuming more RUs in V3 ?

### ISSUE THREAD

**Issue by bartelink** [Azure/azure-cosmos-dotnet-v3#722]
It's unlikely select is broken, but two tests that validate RU consumption consistently fail in V3 but pass in V2. The likelihood is that my port (https://github.com/jet/equinox/pull/144) is making an incorrect mapping from old -> new API

Main suspect is the arguments to the query bit. Would also appreciate any general remarks on the PR

This commit: https://github.com/jet/equinox/commit/9f94b6d88a65f691e0a6e38584dbfb731b44449c reverses a change I had made (upping the tolerance for RU consumption while porting), which makes the tests fail

(I have not done any debugging myself as yet - its entirely possible I'm serializing something wrong, but nonetheless it is passing a significant body of tests)

**Comment by j82w** [comment]
Hi @bartelink,

Can you try setting the max concurrency and max buffered items so they match?

@bchong95 any suggestions?

**Comment by bartelink** [comment]
Tried some combinations without success, but to be honest none of them are speaking to me as being relevant based on the xmldoc?

wrt `PartitionKey`, what is [`<remarks>Only applicable to Item operations</remarks>`](https://github.com/Azure/azure-cosmos-dotnet-v3/blob/0843cae3c252dd49aa8e392623d7eaaed7eb712b/Microsoft.Azure.Cosmos/src/RequestOptions/QueryRequestOptions.cs#L89) intended to convey - i.e. can it link to something or have more detail; I've never encountered the term ? Hm `this.PartitionKey != null && request.ResourceType != ResourceType.Document` - should that become `.Item` ;)

**Comment by j82w** [comment]
Query is supported on Database, Container, and a few other resources. They all share the same QueryRequestOptions. Partition Key is only relevant for query against items, because items are the only resource that has a Partition Key value. Partition Key can have a huge impact on query performance. If the query is restricted to a Partition Key then it only has to go to 1 machine on the backend. If the query does not have a Partition Key then it has to go to all the partitions. The partition key will be extracted from the query when it is possible. There currently is a PR for adding query metrics which will allow you to see what the difference is.

**Comment by bartelink** [comment]
Don't worry, I get all that re partitions etc. I just wasn't aware that QueryRequestOptions was shared in that manner (yuck! NotSupportedException / NotImplementedException is a design smell etc).

**Comment by bartelink** [comment]
Any ideas as to how this could happen ? I showed the before and after code and nothing springs to mind for me. Surely a regression costing RUs like this of interest given it likely has an impact on scalabilty for real apps (assuming its not some parsing anomaly etc)

**Comment by j82w** [comment]
Are both clients running in direct + tcp mode? v2 default to gateway mode + https. V3 defaults to direct + tcp. Direct + tcp is faster, but it's possible it might cost additional RUs if the application is not x64 running on windows because it will need to go to gateway to get the query plan.

**Comment by bartelink** [comment]
@j82w Thanks for the response

My port actually implements mirroring of the V2 default; therefore 
a) the pass on V2 is Gateway mode
b) the fail on V3 is Gateway mode - it is taking more RUs than V2 [in gateway mode]

For fun, [I switched the test suite to run in Direct mode](https://github.com/jet/equinox/pull/144/commits/88f77c56214cbcb786349cb9a159dfc70ed2478d) against V3
a) it continues to fail
b) I'd like to point out that it's pretty ugly to [have to `PackageReference` Azure.Cosmos.Direct](https://github.com/jet/equinox/pull/144/commits/88f77c56214cbcb786349cb9a159dfc70ed2478d#diff-695f944b4fe1c496b82fece6101861abR30) to express a non-default preference - perhaps this should become `bool UseGatewayMode` or something at some point?

The test suite (which passes as long as I remove the RU threshold checks) are pretty thorough, i.e. it's pretty definitely not us making extra API calls etc. Therefore for my money, the suspicion must fall on my porting of the [query logic](https://github.com/jet/equinox/blob/cosmossdk3/src/Equinox.Cosmos/Cosmos.fs#L589); is it possible some flag/default is coming into play?

Calls of stored procs and/or point reads are not raising any alarm bells for me.

Again, my full PR: https://github.com/jet/equinox/pull/144

**Comment by bartelink** [comment]
Hm, some light is appearing. I've [extended another test regarding how responses are batched](https://github.com/jet/equinox/pull/144/commits/37dfe5f9ac6f953c61acb51d6d9a928d7a4f7f9e), which has shown that [QueryRequestOptions.MaxItemCount setting](https://github.com/jet/equinox/blob/cosmossdk3/src/Equinox.Cosmos/Cosmos.fs#L597) is not being honored; I'm requesting:

```fsharp
let qro = new QueryRequestOptions(PartitionKey = Nullable(PartitionKey stream), MaxItemCount=Nullable maxItems)
container.GetItemQueryIterator<Batch>(query, requestOptions = qro)
```

But am seeing many more items being returned - this accounts for the extra charges. However it should be noted that I'm intentionally applying the `MaxItemCount` in order to demonstrate that I can concoct a query such that I can avoid redundant reads (and according charges) when desired.

I've tried also setting the `MaxBufferedItemCount` to the same value (the limit is `1` in some tests and `2` in another)

Can this change in behavior be explained ? If not, can it be mitigated in some alternate manner?

**Comment by bartelink** [comment]
This was finally resolved in [`3.9.0-preview`](https://github.com/Azure/azure-cosmos-dotnet-v3/issues/990#event-3343078993) via #990

NOTE `4.0.0-preview3` still exhibits the discrepancy (and that's likely what I'll target in Equinox's next release)
I'll leave this Issue open until there's a v4 release incorporating these changes that I've validated.

**Comment by j82w** [comment]
Closing this since it is fixed in the latest SDK. V4 SDK will be updated once work to GA it starts.
